File: scala/33.scala
object Solution {
    def search(nums: Array[Int], target: Int): Int = {
      def s(i: Int, j: Int): Int = {
        if (i >= j) return if (nums(i) == target) i else -1
        val mid = i + (j-i) / 2
        if (nums(i) > nums(j)) {
          val l = s(i, mid)
          if (l != -1) return l
          val r = s(mid+1, j)
          if (r != -1) return r
          return -1
        } else {
          if (nums(i) > target || nums(j) < target) return -1
          if (nums(mid) >= target) return s(i, mid)
          else return s(mid+1, j)
        }
      }
      s(0, nums.length-1)
    }
}

File: go/33.go
func search(nums []int, target int) int {
	return s(nums, 0, len(nums)-1, target)
}

func s(nums []int, l, r, target int) int {
	if l > r {
		return -1
	}
	if l == r {
		if nums[l] == target {
			return l
		}
		return -1
	}
	mid := (r-l)/2 + l
	if nums[mid] == target {
		return mid
	}
	if nums[mid] > nums[l] && target < nums[mid] && target >= nums[l] || nums[mid] < nums[r] && (target < nums[mid] || target > nums[r]) {
		return s(nums, l, mid-1, target)
	} else {
		return s(nums, mid+1, r, target)
	}
}